sync15-adapter/src/request.rs

```diff
@@ -0,0 +1,1203 @@
+
+impl Default for InfoConfiguration {
+    #[inline]
+    fn default() -> InfoConfiguration {
+        InfoConfiguration {
+            max_request_bytes: default_max_request_bytes(),
+            max_record_payload_bytes: default_max_record_payload_bytes(),
+            max_post_records: usize::max_value(),
+            max_post_bytes: usize::max_value(),
+            max_total_records: usize::max_value(),
+            max_total_bytes: usize::max_value(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Deserialize)]
+pub struct UploadResult {
+    batch: Option<String>,
+    /// Maps record id => why failed
+    #[serde(default = "HashMap::new")]
+    pub failed: HashMap<String, String>,
+    /// Vec of ids
+    #[serde(default = "Vec::new")]
+    pub success: Vec<String>
+}
+
+// Easier to fake during tests
+#[derive(Debug, Clone)]
+pub struct PostResponse {
+    pub status: StatusCode,
+    pub result: UploadResult, // This is lazy...
+    pub last_modified: ServerTimestamp,
+}
+
+impl PostResponse {
+    pub fn from_response(r: &mut Response) -> Result<PostResponse> {
+        let result: UploadResult = r.json()?;
+        // TODO Can this happen in error cases?
+        let last_modified = r.headers().get::<XLastModified>().map(|h| **h).ok_or_else(||
+            error::unexpected("Server didn't send X-Last-Modified header"))?;
+        let status = r.status();
+        Ok(PostResponse { status, result, last_modified })
+    }
+}
+
+
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub enum BatchState {
+    Unsupported,
+    NoBatch,
+    InBatch(String),
+}
+
+#[derive(Debug)]
+pub struct PostQueue<Post, OnResponse> {
+    poster: Post,
+    // XXX: It's cludgey that this needs to be pub just so we can get at the data it saves...
+    on_response: OnResponse,
+    post_limits: LimitTracker,
+    batch_limits: LimitTracker,
+    max_payload_bytes: usize,
+    max_request_bytes: usize,
+    queued: Vec<u8>,
+    batch: BatchState,
+    last_modified: ServerTimestamp,
+}
+
+pub trait BatchPoster {
+    /// Note: Last argument (reference to the batch poster) is provided for the purposes of testing
+    /// Important: Poster should not report non-success HTTP statuses as errors!!
+    fn post<P, O>(&self,
+            body: &[u8],
+            xius: ServerTimestamp,
+            batch: Option<String>,
+            commit: bool,
+            queue: &PostQueue<P, O>) -> Result<PostResponse>;
+}
+
+// We don't just use a FnMut here since we want to override it in mocking for RefCell<TestType>,
+// which we can't do for FnMut since neither FnMut nor RefCell are defined here. Also, this
+// is somewhat better for documentation.
+pub trait PostResponseHandler {
+    fn handle_response(&mut self, r: PostResponse, mid_batch: bool) -> Result<()>;
+}
+
+
+#[derive(Debug, Clone)]
+pub(crate) struct NormalResponseHandler {
+    pub failed_ids: Vec<String>,
+    pub successful_ids: Vec<String>,
+    pub allow_failed: bool,
```

Comment:
allow_failed is a concept that's really only known to the consumer of this lib, so our API either needs a way for this to be specified, or our result should include enough information for the consumer to detect and abort in this case (I think the latter seems easy enough and the cost of continuing with the batch in this case seems reasonable, as even with these semantics as implemented, we may still have done a commit before we detect a failed record)

Reply:
> we may still have done a commit before we detect a failed record)

This is true on desktop as well, but I believe that it's only possible if the failed record is part of the request that performs the commit. In practice it should be "impossible" because we prevent all the cases where the server is known to reject records before posting.